### PR TITLE
Update armadillo to latest version + support apple M1

### PR DESCRIPTION
**Pull request by quinnj**

### CODE REVIEW

File: A/armadillo/build_tarballs.jl [hunk 1]
```diff
@@ -5,10 +5,10 @@ using BinaryBuilder
 
 
 name = "armadillo"
-version = v"9.850.1"
+version = v"11.2.1"
 sources = [
-    ArchiveSource("http://sourceforge.net/projects/arma/files/armadillo-9.850.1.tar.xz",
-                  "d4c389b9597a5731500ad7a2656c11a6031757aaaadbcafdea5cc8ac0fd2c01f")
+    ArchiveSource("http://sourceforge.net/projects/arma/files/armadillo-11.2.1.tar.xz",
```
Comment (giordano) [inline A/armadillo/build_tarballs.jl:10]:
```suggestion
    ArchiveSource("http://sourceforge.net/projects/arma/files/armadillo-$(version).tar.xz",
```

File: A/armadillo/build_tarballs.jl [hunk 2]
```diff
@@ -20,31 +20,11 @@ FLAGS=(-DCMAKE_TOOLCHAIN_FILE=${CMAKE_TARGET_TOOLCHAIN}
...
-        SYMB_DEFS+=("-D${sym}=${sym}_64")
-    done
-
-    export CXXFLAGS="${SYMB_DEFS[@]}"
-
-    # Force the configuration parameter ARMA_BLAS_LONG to be true, as in our
-    # setting 64-bit systems are going to need a 64-bit integer to be used for
-    # Armadillo's `blas_int` type.
-    sed -i 's|// #define ARMA_BLAS_LONG$|#define ARMA_BLAS_LONG|' ../include/armadillo_bits/config.hpp.cmake
-    cat ../include/armadillo_bits/config.hpp.cmake
```
Comment (rcurtin) [inline A/armadillo/build_tarballs.jl:47]: I'm doubtful that this should be removed.  Armadillo is essentially a header-only C++ wrapper around LAPACK and BLAS, and it has to reference the correct symbols from OpenBLAS.  Unless Yggdrasil has changed how it distributes OpenBLAS on OS X, I strongly suspect that removing this is incorrect.

File: A/armadillo/build_tarballs.jl [hunk 3]
```diff
@@ -20,31 +20,11 @@ FLAGS=(-DCMAKE_TOOLCHAIN_FILE=${CMAKE_TARGET_TOOLCHAIN}
...
-    FLAGS+=(-DLAPACK_LIBRARY="${libdir}/libopenblas64_.${dlext}")
+if [[ "${nbits}" == 64 ]]; then
 
-    SYMB_DEFS=()
-    for sym in sasum dasum snrm2 dnrm2 sdot ddot sgemv dgemv cgemv zgemv sgemm dgemm cgemm zgemm ssyrk dsyrk cherk zherk; do
-        SYMB_DEFS+=("-D${sym}=${sym}_64")
-    done
     if [[ "${target}" == *-apple-* ]] || [[ "${target}" == *-mingw* ]]; then
-        FLAGS+=(-DALLOW_OPENBLAS_MACOS=ON)
+        FLAGS+=(-DALLOW_BLAS_LAPACK_MACOS=ON)
```
Comment (rcurtin) [inline A/armadillo/build_tarballs.jl:26]: I don't think this needs to change either.  From the Armadillo README:

```
    - On macOS, to enable the detection of OpenBLAS,
      use the additional ALLOW_OPENBLAS_MACOS option when running cmake:

      cmake -DALLOW_OPENBLAS_MACOS=ON .

      Note: depending on your installation, OpenBLAS may masquerade as standard BLAS.
      To detect standard BLAS and LAPACK, use the ALLOW_BLAS_LAPACK_MACOS option:

      cmake -DALLOW_BLAS_LAPACK_MACOS=ON .
```

Unless OpenBLAS is now distributed differently, I suspect that this change is not needed.

**Review by rcurtin — CHANGES_REQUESTED** [review]
I have some concerns about these changes.  Specifically one of the things that is difficult here is that Armadillo is a header-only library that, at configuration time, detects what libraries are available and sets its installed configuration (in `armadillo_bits/config.hpp`) accordingly.  But, if you then try to link against Armadillo, you will have all kinds of symbol not found errors if things weren't set up correctly.

I see that more recent versions of Armadillo have a `BUILD_SMOKE_TESTS` CMake option.  Perhaps it is worthwhile to try and enable that, and ensure that it builds a program that links against Armadillo successfully?

Alternately, you might try testing a dependent package (I think mlpack may be the only one) to make sure it links correctly.

**Comment by rcurtin** [comment]
> I see that more recent versions of Armadillo have a `BUILD_SMOKE_TESTS` CMake option. Perhaps it is worthwhile to try and enable that, and ensure that it builds a program that links against Armadillo successfully?

Sorry, I am incorrect here!  I am now looking at the build logs and see that this is indeed being tested.  Let me now skim the rest of the build logs and see if I can understand more closely what is going on here.  Hopefully given what I explained above you can at least see why the changes gave me a little heartburn. :)

**Comment by rcurtin** [comment]
In the more recent builds, Armadillo is now configuring *without* LAPACK or BLAS support:

```
[17:25:21] -- FlexiBLAS_FOUND = NO
[17:25:21] --       MKL_FOUND = NO
[17:25:21] --  OpenBLAS_FOUND = NO
[17:25:21] --     ATLAS_FOUND = NO
[17:25:21] --      BLAS_FOUND = NO
[17:25:21] --    LAPACK_FOUND = NO
```

i.e., it is not finding OpenBLAS.  That will definitely be a problem for downstream packages, although the smoke tests will still successfully build in this case.  I did not check every build, but at least for `aarch64-linux-musl` and `x86_64-linux-musl` OpenBLAS is not being detected.  (It is correct for `i686-linux-musl`, but there have been no changes there, so that is what I'd expect.)

Looking more closely at, e.g., [this build](https://dev.azure.com/JuliaPackaging/Yggdrasil/_build/results?buildId=20407&view=logs&jobId=a80b813d-4353-593b-f910-8018f3f76c79&j=8e4e1a87-e815-52c1-9bd7-3976c938457a&t=4885fdfd-fd3d-5436-8b54-16a16d112f49), the linking errors are:

```
[18:38:09] ../libarmadillo.so.11.2.1: undefined reference to `dpstrf_'
[18:38:09] ../libarmadillo.so.11.2.1: undefined reference to `dgeqp3_'
[18:38:09] ../libarmadillo.so.11.2.1: undefined reference to `cpstrf_'
[18:38:09] ../libarmadillo.so.11.2.1: undefined reference to `zpstrf_'
[18:38:09] ../libarmadillo.so.11.2.1: undefined reference to `spstrf_'
[18:38:09] ../libarmadillo.so.11.2.1: undefined reference to `sgeqp3_'
[18:38:09] ../libarmadillo.so.11.2.1: undefined reference to `zgeqp3_'
[18:38:09] ../libarmadillo.so.11.2.1: undefined reference to `cgeqp3_'
```

@giordano has there been a change in the naming conventions for OpenBLAS on 64-bit systems?  Alternately, I wonder if those particular symbols just did not appear in the big list of symbols that we were rewriting in `build_tarballs.jl`, and just need to be added.  For example, I don't see `cpstrf` on the list.  Maybe newer Armadillo uses those symbols, but older Armadillo didn't (hence their lack of inclusion on previous lists).  @quinnj do you want to try that and see what happens?

**Comment by quinnj** [comment]
@rcurtin, thanks for the pointers and suggestions! Yes, adding those symbols to the list seems to unblock things at least locally. I've pushed the change here (and removed my previous commit which removed a bunch of things), so let's see what CI thinks.

**Comment by quinnj** [comment]
Ok, now we're just seeing a powerpc error:
```
[20:12:26] [100%] Linking CXX executable smoke_test
[20:12:26] cd /workspace/srcdir/armadillo-11.2.1/build/tests1 && /usr/bin/cmake -E cmake_link_script CMakeFiles/smoke_test.dir/link.txt --verbose=true
[20:12:26] /opt/bin/powerpc64le-linux-gnu-libgfortran3-cxx03/powerpc64le-linux-gnu-g++ --sysroot=/opt/powerpc64le-linux-gnu/powerpc64le-linux-gnu/sys-root/ -Dsasum=sasum_64 -Ddasum=dasum_64 -Dsnrm2=snrm2_64 -Ddnrm2=dnrm2_64 -Dsdot=sdot_64 -Dddot=ddot_64 -Dsgemv=sgemv_64 -Ddgemv=dgemv_64 -Dcgemv=cgemv_64 -Dzgemv=zgemv_64 -Dsgemm=sgemm_64 -Ddgemm=dgemm_64 -Dcgemm=cgemm_64 -Dzgemm=zgemm_64 -Dssyrk=ssyrk_64 -Ddsyrk=dsyrk_64 -Dcherk=cherk_64 -Dzherk=zherk_64 -Dcgbcon=cgbcon_64 -Dcgbsv=cgbsv_64 -Dcgbsvx=cgbsvx_64 -Dcgbtrf=cgbtrf_64 -Dcgbtrs=cgbtrs_64 -Dcgecon=cgecon_64 -Dcgees=cgees_64 -Dcgeev=cgeev_64 -Dcgeevx=cgeevx_64 -Dcgehrd=cgehrd_64 -Dcgels=cgels_64 -Dcgelsd=cgelsd_64 -Dcgemm=cgemm_64 -Dcgemv=cgemv_64 -Dcgeqp3=cgeqp3_64 -Dcgeqrf=cgeqrf_64 -Dcgesdd=cgesdd_64 -Dcgesv=cgesv_64 -Dcgesvd=cgesvd_64 -Dcgesvx=cgesvx_64 -Dcgetrf=cgetrf_64 -Dcgetri=cgetri_64 -Dcgetrs=cgetrs_64 -Dcgges=cgges_64 -Dcggev=cggev_64 -Dcgtsv=cgtsv_64 -Dcgtsvx=cgtsvx_64 -Dcheev=cheev_64 -Dcheevd=cheevd_64 -Dcherk=cherk_64 -Dclangb=clangb_64 -Dclange=clange_64 -Dclanhe=clanhe_64 -Dclansy=clansy_64 -Dcpbtrf=cpbtrf_64 -Dcpocon=cpocon_64 -Dcposv=cposv_64 -Dcposvx=cposvx_64 -Dcpotrf=cpotrf_64 -Dcpotri=cpotri_64 -Dcpotrs=cpotrs_64 -Dcpstrf=cpstrf_64 -Dctrcon=ctrcon_64 -Dctrsyl=ctrsyl_64 -Dctrtri=ctrtri_64 -Dctrtrs=ctrtrs_64 -Dcungqr=cungqr_64 -Ddasum=dasum_64 -Dddot=ddot_64 -Ddgbcon=dgbcon_64 -Ddgbsv=dgbsv_64 -Ddgbsvx=dgbsvx_64 -Ddgbtrf=dgbtrf_64 -Ddgbtrs=dgbtrs_64 -Ddgecon=dgecon_64 -Ddgees=dgees_64 -Ddgeev=dgeev_64 -Ddgeevx=dgeevx_64 -Ddgehrd=dgehrd_64 -Ddgels=dgels_64 -Ddgelsd=dgelsd_64 -Ddgemm=dgemm_64 -Ddgemv=dgemv_64 -Ddgeqp3=dgeqp3_64 -Ddgeqrf=dgeqrf_64 -Ddgesdd=dgesdd_64 -Ddgesv=dgesv_64 -Ddgesvd=dgesvd_64 -Ddgesvx=dgesvx_64 -Ddgetrf=dgetrf_64 -Ddgetri=dgetri_64 -Ddgetrs=dgetrs_64 -Ddgges=dgges_64 -Ddggev=dggev_64 -Ddgtsv=dgtsv_64 -Ddgtsvx=dgtsvx_64 -Ddlahqr=dlahqr_64 -Ddlangb=dlangb_64 -Ddlange=dlange_64 -Ddlansy=dlansy_64 -Ddlarnv=dlarnv_64 -Ddnrm2=dnrm2_64 -Ddorgqr=dorgqr_64 -Ddpbtrf=dpbtrf_64 -Ddpocon=dpocon_64 -Ddposv=dposv_64 -Ddposvx=dposvx_64 -Ddpotrf=dpotrf_64 -Ddpotri=dpotri_64 -Ddpotrs=dpotrs_64 -Ddpstrf=dpstrf_64 -Ddstedc=dstedc_64 -Ddsyev=dsyev_64 -Ddsyevd=dsyevd_64 -Ddsyrk=dsyrk_64 -Ddtrcon=dtrcon_64 -Ddtrevc=dtrevc_64 -Ddtrsyl=dtrsyl_64 -Ddtrtri=dtrtri_64 -Ddtrtrs=dtrtrs_64 -Dilaenv=ilaenv_64 -Dsasum=sasum_64 -Dsdot=sdot_64 -Dsgbcon=sgbcon_64 -Dsgbsv=sgbsv_64 -Dsgbsvx=sgbsvx_64 -Dsgbtrf=sgbtrf_64 -Dsgbtrs=sgbtrs_64 -Dsgecon=sgecon_64 -Dsgees=sgees_64 -Dsgeev=sgeev_64 -Dsgeevx=sgeevx_64 -Dsgehrd=sgehrd_64 -Dsgels=sgels_64 -Dsgelsd=sgelsd_64 -Dsgemm=sgemm_64 -Dsgemv=sgemv_64 -Dsgeqrf=sgeqrf_64 -Dsgeqp3=sgeqp3_64 -Dsgesdd=sgesdd_64 -Dsgesv=sgesv_64 -Dsgesvd=sgesvd_64 -Dsgesvx=sgesvx_64 -Dsgetrf=sgetrf_64 -Dsgetri=sgetri_64 -Dsgetrs=sgetrs_64 -Dsgges=sgges_64 -Dsggev=sggev_64 -Dsgtsv=sgtsv_64 -Dsgtsvx=sgtsvx_64 -Dslahqr=slahqr_64 -Dslangb=slangb_64 -Dslange=slange_64 -Dslansy=slansy_64 -Dslarnv=slarnv_64 -Dsnrm2=snrm2_64 -Dsorgqr=sorgqr_64 -Dspbtrf=spbtrf_64 -Dspocon=spocon_64 -Dsposv=sposv_64 -Dsposvx=sposvx_64 -Dspotrf=spotrf_64 -Dspotri=spotri_64 -Dspotrs=spotrs_64 -Dspstrf=spstrf_64 -Dsstedc=sstedc_64 -Dssyev=ssyev_64 -Dssyevd=ssyevd_64 -Dssyrk=ssyrk_64 -Dstrcon=strcon_64 -Dstrevc=strevc_64 -Dstrsyl=strsyl_64 -Dstrtri=strtri_64 -Dstrtrs=strtrs_64 -Dzgbcon=zgbcon_64 -Dzgbsv=zgbsv_64 -Dzgbsvx=zgbsvx_64 -Dzgbtrf=zgbtrf_64 -Dzgbtrs=zgbtrs_64 -Dzgecon=zgecon_64 -Dzgees=zgees_64 -Dzgeev=zgeev_64 -Dzgeevx=zgeevx_64 -Dzgehrd=zgehrd_64 -Dzgels=zgels_64 -Dzgelsd=zgelsd_64 -Dzgemm=zgemm_64 -Dzgemv=zgemv_64 -Dzgeqp3=zgeqp3_64 -Dzgeqrf=zgeqrf_64 -Dzgesdd=zgesdd_64 -Dzgesv=zgesv_64 -Dzgesvd=zgesvd_64 -Dzgesvx=zgesvx_64 -Dzgetrf=zgetrf_64 -Dzgetri=zgetri_64 -Dzgetrs=zgetrs_64 -Dzgges=zgges_64 -Dzggev=zggev_64 -Dzgtsv=zgtsv_64 -Dzgtsvx=zgtsvx_64 -Dzheev=zheev_64 -Dzheevd=zheevd_64 -Dzherk=zherk_64 -Dzlangb=zlangb_64 -Dzlange=zlange_64 -Dzlanhe=zlanhe_64 -Dzlansy=zlansy_64 -Dzpbtrf=zpbtrf_64 -Dzpocon=zpocon_64 -Dzposv=zposv_64 -Dzposvx=zposvx_64 -Dzpotrf=zpotrf_64 -Dzpotri=zpotri_64 -Dzpotrs=zpotrs_64 -Dzpstrf=zpstrf_64 -Dztrcon=ztrcon_64 -Dztrsyl=ztrsyl_64 -Dztrtri=ztrtri_64 -Dztrtrs=ztrtrs_64 -Dzungqr=zungqr_64 -O2 -O3 -DNDEBUG -rdynamic CMakeFiles/smoke_test.dir/smoke_test.cpp.o -o smoke_test  -Wl,-rpath,/workspace/srcdir/armadillo-11.2.1/build:/workspace/destdir/lib ../libarmadillo.so.11.2.1 /workspace/destdir/lib/libopenblas64_.so 
[20:12:27] /workspace/destdir/lib/libopenblas64_.so: undefined reference to `GOMP_parallel@GOMP_4.0'
[20:12:27] /workspace/destdir/lib/libopenblas64_.so: undefined reference to `omp_get_num_places@OMP_4.5'
[20:12:27] collect2: error: ld returned 1 exit status

```

**Comment by rcurtin** [comment]
Ok, awesome, I guess that there has not been a change to OpenBLAS packaging, we just have some new symbols.

The PPC error... the compiler probably needs `-fopenmp` because OpenBLAS appears to be compiled with it, but Armadillo is agnostic to it.  I think you might be able to get it to work on that platform with something like `-DCMAKE_CXX_FLAGS="-fopenmp"` (that's a little bit abusive but I think it might work).  Give it a shot and see what happens?

I think Armadillo's smoke test CMake configuration is incorrect in this case, so we probably need to work around it.

**Comment by quinnj** [comment]
Hmmm, I'm not sure I quite added the flag correctly, but the latest build does error _differently_ 😅 . But we're now seeing lots of undefined reference errors.

**Comment by rcurtin** [comment]
CMake is such fun!  If `CMAKE_CXX_FLAGS` is set, then `CXXFLAGS` is ignored:
https://cmake.org/cmake/help/latest/envvar/CXXFLAGS.html

So I guess just append `-fopenmp` to `CXXFLAGS` once the big list of symbols is assembled, maybe that will work?

**Comment by quinnj** [comment]
Ok, tried that, but it didnt' seem to change anything.

File: A/armadillo/build_tarballs.jl [hunk 4]
```diff
@@ -33,15 +33,15 @@ if [[ "${nbits}" == 64 ]]; then
...
         SYMB_DEFS+=("-D${sym}=${sym}_64")
     done
 
     export CXXFLAGS="${SYMB_DEFS[@]}"
+    if [[ "${target}" == powerpc64le-* ]]; then
+        CXXFLAGS+=(-fopenmp)
+    fi
```
Comment (rcurtin) [inline A/armadillo/build_tarballs.jl:44]: I think you need to put this before the `export`, which flattens the array into a string.  I'm not 100% sure on that, though.

Comment (rcurtin) [reply A/armadillo/build_tarballs.jl:43]: Sorry, I think my suggestion was bad... maybe it suffices to just do this:

```
if [[ "${target}" == powerpc64le-* ]]; then
    export CXXFLAGS="${SYMB_DEFS[@]} -fopenmp"
else
    export CXXFLAGS="${SYMB_DEFS[@]}"
fi
```

(You may want to double-check that, I didn't test it or anything.  Maybe printing `$CXXFLAGS` to ensure it's correct is worthwhile.)

**Comment by rcurtin** [comment]
It's also possible that `-fopenmp` needs to be a part of `LDFLAGS`; I'm not 100% sure.

**Comment by giordano** [comment]
I never remember why, but we simply need to use GCC 6 for using OpenMP with PowerPC.

**Comment by quinnj** [comment]
Thank you @giordano!! ❤️

**Pull request merged**